### PR TITLE
Context Manager and timeout on `self.wpipe.close`

### PR DESCRIPTION
**Pull request by romanlevin**
This is kind of rough at the moment (at very least, the timeout needs to be configurable), but I was wondering if you'd be interested in a cleaned up version of this.

### CODE REVIEW

**Comment by romanlevin** [comment]
Pinging @lordmauve

**Comment by lordmauve** [comment]
Hi, sorry for the delay.

I think it might be a better idea for `close()` to simply signal the process in the default case, rather than waiting.

I wouldn't be averse to having the stronger version - wait for process to die, or kill it  - with something like close(force=True).

File: chopsticks/tunnel.py [hunk 1]
```diff
@@ -387,9 +387,26 @@ def cmd_args(self):
...
-    def __del__(self):
+    def close(self):
         self.wpipe.close()  # Terminate child
-        self.proc.wait()
+        if self.proc.poll() is not None:
+            # subprocess is already dead
+            return
+        try:
+            self.proc.terminate()
+            self.proc.wait(timeout=5)
```
Comment (lordmauve) [inline chopsticks/tunnel.py]: The timeout parameter isn't Python 2.7-compatible.

**Comment by romanlevin** [comment]
@lordmauve I replaced `timeout` with a polling loop. Do you still think it should just not wait by default?

**Comment by romanlevin** [comment]
@lordmauve I've tested with both Python 2.7.13 (wheren `Popen.stdin` is a file descriptor) and with Python 3.6.1 (where it is an `io.BufferedWriter`), and both allow multiple calls to `close()`.

Any subsequent calls are noops.

**Comment by lordmauve** [comment]
Ok, then!

**Pull request merged**